pkg/client/client_v2.go

```diff
@@ -0,0 +1,541 @@
+		token:           token,
+		versionInfo:     versionInfo,
+		kickChan:        make(chan struct{}, 1),
+		errChan:         make(chan error),
+		unitsChan:       make(chan UnitChanged),
+		minCheckTimeout: CheckinMinimumTimeout,
+	}
+}
+
+// Start starts the connection to Elastic Agent.
+func (c *clientV2) Start(ctx context.Context) error {
+	c.ctx, c.cancel = context.WithCancel(ctx)
+	conn, err := grpc.DialContext(ctx, c.target, c.opts...)
+	if err != nil {
+		return err
+	}
+	c.client = proto.NewElasticAgentClient(conn)
+	c.storeClient = proto.NewElasticAgentStoreClient(conn)
+	c.artifactClient = proto.NewElasticAgentArtifactClient(conn)
+	c.logClient = proto.NewElasticAgentLogClient(conn)
+	c.startCheckin()
+	c.startActions()
+	return nil
+}
+
+// Stop stops the connection to Elastic Agent.
+func (c *clientV2) Stop() {
+	if c.cancel != nil {
+		c.cancel()
+		c.wg.Wait()
+		c.ctx = nil
+		c.cancel = nil
+	}
+}
+
+// UnitChanges returns channel client send unit change notifications to.
+func (c *clientV2) UnitChanges() <-chan UnitChanged {
+	return c.unitsChan
+}
+
+// Errors returns channel of errors that occurred during communication.
+func (c *clientV2) Errors() <-chan error {
+	return c.errChan
+}
+
+// Artifacts returns the artifacts client.
+func (c *clientV2) Artifacts() ArtifactsClient {
+	return &artifactsClient{c}
+}
+
+// startCheckin starts the go routines to send and receive check-ins
+//
+// This starts 3 go routines to manage the check-in bi-directional stream. The first
+// go routine starts the stream then starts one go routine to receive messages and
+// another go routine to send messages. The first go routine then blocks waiting on
+// the receive and send to finish, then restarts the stream or exits if the context
+// has been cancelled.
+func (c *clientV2) startCheckin() {
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		for {
+			select {
+			case <-c.ctx.Done():
+				// stopped
+				return
+			default:
+			}
+
+			c.checkinRoundTrip()
+		}
+	}()
+}
+
+func (c *clientV2) checkinRoundTrip() {
+	checkinCtx, checkinCancel := context.WithCancel(c.ctx)
+	defer checkinCancel()
+
+	checkinClient, err := c.client.CheckinV2(checkinCtx)
+	if err != nil {
+		c.errChan <- err
+		return
+	}
+
+	// ensure first checkin include version information
+	c.versionInfoSent = false
+
+	var checkinRead sync.WaitGroup
+	var checkinWrite sync.WaitGroup
+	done := make(chan bool)
+
+	// expected state check-ins
+	checkinRead.Add(1)
+	go func() {
+		defer checkinRead.Done()
+		for {
+			expected, err := checkinClient.Recv()
+			if err != nil {
+				if err != io.EOF {
+					c.errChan <- err
+				}
+				close(done)
+				return
+			}
+			c.syncUnits(expected)
+		}
+	}()
+
+	// observed state check-ins
+	checkinWrite.Add(1)
+	go func() {
+		defer checkinWrite.Done()
+
+		if err := c.sendObserved(checkinClient); err != nil {
+			if err != io.EOF {
+				c.errChan <- err
+			}
+			return
+		}
+
+		t := time.NewTicker(c.minCheckTimeout)
```

Comment:
Based on our morning discussion this would behavior weirdlky with sleep or hibernate?

Reply:
After looking at the golang I do not believe this to be an issue. The ticker will keep ticking, on hibernate it will just pickup where it left off because it uses the monoclock. The ticker also handles a slow reader so it never adds more than one to the channel so until the next is removed from the channel it will not add another one.